File: programming 1/lab 5/src/managers/CollectionManager.java
package managers;

import exceptions.MissingFileException;
import models.StudyGroup;
import utils.ObjectBuilder;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

/**
 * Менеджер, оперирующий коллекцией
 */

public class CollectionManager {
    private final PriorityQueue<StudyGroup> collection;
    private final DumpManager dumpManager;
    private final String loadFile;
    private LocalDateTime initializationTime;
    private LocalDateTime lastSaveTime;
    private PriorityQueue<Long> availableIds = new PriorityQueue<>();
    private Set<Long> usedIds = new HashSet<>();
    private Long nextId = 1L;

    public CollectionManager(PriorityQueue<StudyGroup> collection, String loadFile) {
        this.collection = collection;
        this.loadFile = loadFile;
        this.dumpManager = new DumpManager(collection);
        this.lastSaveTime = null;

        load();
        this.initializationTime = LocalDateTime.now();
        for (StudyGroup object : collection) {
            Long id = object.getId();
            usedIds.add(id);
            if (id.equals(nextId)) {
                nextId = nextId + 1;
            } else {
                availableIds.add(id);
            }
        }
    }

    /**
     * Проверяет, свободен ли указанный идентификатор в коллекции.
     * @param id идентификатор для проверки.
     * @return true, если идентификатор свободен; false в противном случае.
     */
    public boolean idIsFree(Long id) {
        return !(usedIds.contains(id));
    }


    /**
     * @return свободный id.
     */
    public Long getId() {
        Long id;
        if (!availableIds.isEmpty()) {
            id = availableIds.poll();
        } else {
            id = nextId;
            nextId += 1;
            usedIds.add(id);
        }
        return id;
    }

    /**
     * Добавляет новый элемент в коллекцию.
     */
    public void add() {
        collection.add(ObjectBuilder.create(getId()));
        System.out.println("Элемент успешно добавлен!");
    }

    /**
     * Добавляет новый элемент в коллекцию на основе данных из файла.
     * @param reader для чтения данных из файла.
     */
    public void add(BufferedReader reader) {
        StudyGroup studyGroup = ObjectBuilder.create(getId(), reader);
        if (studyGroup != null) {
            collection.add(studyGroup);
            System.out.println("Элемент успешно добавлен!");
        }
    }

    /**
     * Выводит содержимое коллекции.
     */

    public void show() {
        if (collection.isEmpty()) System.out.println("Коллекция пуста!");
        else {
            PriorityQueue<StudyGroup> tempCollection = new PriorityQueue<>(collection);
            while (!tempCollection.isEmpty()) {
                System.out.println(tempCollection.poll());
            }
        }
    }

    /**
     * Очищает коллекцию.
     */
    public void clear() {
        collection.clear();
        System.out.println("Коллекция пуста!");
    }

    /**
     * Выводит первый элемент коллекции.
     */
    public void head() {
        if (collection.isEmpty()) System.out.println("Коллекция пуста!");
        else System.out.println(collection.peek());
    }

    /**
     * Удаляет элемент из коллекции по указанному идентификатору.
     * @param id идентификатор элемента для удаления.
     */
    public void remove(Long id) {
        if (idIsFree(id)) {
            System.err.printf("Элемент с id=%s не найден!\n", id);
            return;
        }
        availableIds.add(id);
        for (StudyGroup studyGroup: collection) {
            if (studyGroup.getId().equals(id)) {
                collection.remove(studyGroup);
                System.out.println("Элемент успешно удалён!");
                break;
            }
        }
    }

    /**
     * Обновляет элемент из коллекции по указанному идентификатору.
     * @param id идентификатор элемента для обновления.
     */
    public void update(Long id) {
        if (idIsFree(id)) {
            System.err.printf("Элемент с id=%s не найден!\n", id);
            return;
        }
        for (StudyGroup studyGroup: collection) {
            if (studyGroup.getId().equals(id)) {
                collection.remove(studyGroup);
                break;
            }
        }
        collection.add(ObjectBuilder.create(id));
        System.out.println("Элемент успешно обновлён!");
    }

    /**
     * Обновляет элемент из коллекции по указанному идентификатору и берёт данные для обновления из файла.
     * @param id идентификатор элемента для удаления.
     * @param reader для чтения данных из файла.
     */
    public void update(BufferedReader reader, Long id) {
        if (idIsFree(id)) {
            System.err.printf("Элемент с id=%s не найден!\n", id);
            return;
        }
        for (StudyGroup studyGroup: collection) {
            if (studyGroup.getId().equals(id)) {
                collection.remove(studyGroup);
                break;
            }
        }
        collection.add(ObjectBuilder.create(id, reader));
        System.out.println("Элемент c id = "  + id + " успешно обновлён!");
    }

    /**
     * Удаляет все элементы, превышающие заданный.
     */
    public void removeGreater() {
        if (collection.isEmpty()) System.out.println("Коллекция пуста!");
        else {
            StudyGroup compareStudyGroup = ObjectBuilder.create(Long.valueOf(0));
            for (StudyGroup studyGroup : collection) {
                if (collection.comparator().compare(studyGroup, compareStudyGroup) > 0) {
                    availableIds.add(studyGroup.getId());
                    collection.remove(studyGroup);
                }
            }
            System.out.println("Элемент(ы) успешно удален(ы)!");
        }
    }

    /**
     * Удаляет все элементы, превышающие тот, чьи данные берёт из файла.
     * @param reader для чтения данных из файла.
     */
    public void removeGreater(BufferedReader reader) {
        if (collection.isEmpty()) System.out.println("Коллекция пуста!");
        else {
            StudyGroup compareStudyGroup = ObjectBuilder.create(Long.valueOf(0), reader);
            for (StudyGroup studyGroup : collection) {
                if (collection.comparator().compare(studyGroup, compareStudyGroup) > 0) {
                    availableIds.add(studyGroup.getId());
                    collection.remove(studyGroup);
                }
            }
            System.out.println("Элемент(ы) успешно удален(ы)!");
        }
    }

    /**
     * Выводит количество элементов, значение поля studentsCount которых меньше заданного.
     * @param studentsCount количество студентов.
     */
    public void countLess(long studentsCount) {
        int count = 0;
        for (StudyGroup studyGroup: collection) {
            if (studyGroup.getStudentsCount() < studentsCount) {
                count += 1;
            }
        }
        System.out.println(count);
    }

    /**
     * Выводит элементы коллекции в порядке убывания.
     */
    public void printDescending() {
        if (collection.isEmpty()) System.out.println("Коллекция пуста!");
        else {
            PriorityQueue<StudyGroup> tempCollection = new PriorityQueue<>(collection);

            List<StudyGroup> listCollection = new ArrayList<>();
            while (!tempCollection.isEmpty()) {
                listCollection.add(tempCollection.poll());
            }

            for (int i = listCollection.size() - 1; i >= 0; i--) {
                System.out.println(listCollection.get(i));
            }
        }
    }

    /**
     * Выводит значения поля formOfEducation всех элементов в порядке возрастания.
     */
    public void printFieldAscending() {
        if (collection.isEmpty()) System.out.println("Коллекция пуста!");
        else {
            PriorityQueue<StudyGroup> tempCollection = new PriorityQueue<>(collection);
            while (!tempCollection.isEmpty()) {
                System.out.println(tempCollection.poll().getFormOfEducation());
            }
        }
    }

    /**
     * Выводит информацию о коллекции.
     */

    public void info() {
        System.out.println("Тип коллекции: PriorityQueue");
        System.out.println("Тип элементов: StudyGroup");
        System.out.println("Количество элементов: " + collection.size());
        System.out.println("Дата инициализации: " + initializationTime.format(DateTimeFormatter.ofPattern("dd.MM.yyyy")));
        if (lastSaveTime == null) System.out.println("В текущей сессии сохранение коллекции ещё не осуществлялось.");
        else System.out.println("Дата последнего сохранения: " + lastSaveTime.format(DateTimeFormatter.ofPattern("dd.MM.yyyy")));
    }

    /**
     * Загружает элементы коллекции из файла, который передаётся из аргумента командной строки.
     */
    public void load() {
        try {
            dumpManager.loadCollection(loadFile);
        } catch (MissingFileException e) {
            System.err.println(e.getMessage());
        } catch (FileNotFoundException e) {
            System.err.println("Файл " + loadFile + " не найден.");
        } catch (SecurityException e) {
            System.err.println("Нет прав доступа к файлу " + loadFile);
        } catch (Exception e) {
            System.err.println("Ошибка загрузки коллекции из файла" + loadFile + ": " + e.getMessage());
        }
    }

    /**
     * Сохраняет элементы коллекции в файл, который передаётся из аргумента командной строки.
     */
    public void save() {
        try {
            dumpManager.saveCollection(loadFile);
            lastSaveTime = LocalDateTime.now();
            System.out.println("Коллекция успешно сохранена!");
        } catch (MissingFileException e) {
            System.err.println(e.getMessage());
        } catch (SecurityException e) {
            System.err.println("Нет прав доступа к файлу " + loadFile);
        } catch (Exception e) {
            System.err.println("Ошибка сохранения коллекция в файл" + loadFile + ": " + e.getMessage());
        }
    }
}

File: programming 1/lab 7/src/main/java/server/managers/CollectionManager.kt
package server.managers

import common.communication.Response
import common.communication.ResponseStatus
import common.models.StudyGroup
import common.utils.ObjectBuilder
import server.network.Server
import server.utils.HistoryWriter
import java.io.BufferedReader
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*
import java.util.concurrent.locks.ReentrantReadWriteLock

/**
 * Менеджер, оперирующий коллекцией
 */
class CollectionManager(private val dataBaseManager: DataBaseManager, private val threadPoolManager: ThreadPoolManager) {
    private val locker = ReentrantReadWriteLock()
    private val collection: PriorityQueue<StudyGroup> = dataBaseManager.load()
    private val initializationTime: LocalDateTime = LocalDateTime.now()

    /**
     * Добавляет новый элемент в коллекцию.
     */
    fun add(): Response {
        threadPoolManager.responseSenderExecutor.submit {
            Server.sendResponse(Response(ResponseStatus.ASK_OBJECT, "Введите данные добавляемого объекта!"))
        }
        val studyGroup = threadPoolManager.objectQueue.take()

        HistoryWriter.writeObject(studyGroup)
        locker.writeLock().lock()
        try {
            studyGroup.id = dataBaseManager.add(studyGroup)
            if (studyGroup.id > 0) {
                collection.add(studyGroup)
                return Response(ResponseStatus.SUCCESS, "Элемент успешно добавлен!\n")
            }
        } finally {
            locker.writeLock().unlock()
        }
        return Response(ResponseStatus.ERROR, "Произошла ошибка при добавлении элемента в базу данных!")
    }

    /**
     * Добавляет новый элемент в коллекцию на основе данных из файла.
     * @param reader для чтения данных из файла.
     */
    fun add(reader: BufferedReader): Response {
        val studyGroup = ObjectBuilder.create(reader)
        HistoryWriter.writeObject(studyGroup)

        locker.writeLock().lock()
        try {
            studyGroup.id = dataBaseManager.add(studyGroup)
            if (studyGroup.id > 0) {
                collection.add(studyGroup)
                return Response(ResponseStatus.INTERMEDIATE_SUCCESS, "Элемент успешно добавлен!\n")
            }
        } finally {
            locker.writeLock().unlock()
        }
        return Response(ResponseStatus.INTERMEDIATE_ERROR, "Произошла ошибка при добавлении элемента в базу данных!")
    }


    /**
     * Выводит содержимое коллекции.
     */
    fun show(): Response {
        var descr = ""
        locker.readLock().lock()
        try {
            if (collection.isEmpty()) descr = "Коллекция пуста!"
            collection.stream().sorted().forEach { studyGroup: StudyGroup ->
                val future = threadPoolManager.responseSenderExecutor.submit {
                    Server.sendResponse(Response(ResponseStatus.INTERMEDIATE_SUCCESS, studyGroup.toString()))
                }
                future.get()
            }
        } finally {
            locker.readLock().unlock()
        }
        return Response(ResponseStatus.SUCCESS, descr)
    }

    /**
     * Очищает коллекцию.
     */
    fun clear(): Response {
        locker.writeLock().lock()
        try {
            dataBaseManager.clear()
            collection.clear()
        } finally {
            locker.writeLock().unlock()
        }
        return Response(ResponseStatus.SUCCESS, "Коллекция пуста!")
    }

    /**
     * Выводит первый элемент коллекции.
     */
    fun head(): Response {
        locker.readLock().lock()
        try {
            return if (collection.isEmpty()) Response(
                ResponseStatus.SUCCESS,
                "Коллекция пуста!"
            )
            else Response(ResponseStatus.SUCCESS, collection.peek().toString() + "\n")
        } finally {
            locker.readLock().unlock()
        }
    }

    /**
     * Удаляет элемент из коллекции по указанному идентификатору.
     * @param id идентификатор элемента для удаления.
     */
    fun remove(id: Long): Response {
        if (dataBaseManager.exist(id)) {
            if (dataBaseManager.checkCreatorId(id)) {
                locker.writeLock().lock()
                try {
                    if (dataBaseManager.remove(id)) {
                        val deleteGroup =
                            collection.stream().filter { studyGroup: StudyGroup -> studyGroup.id == id }.findFirst()
                                .get()
                        collection.remove(deleteGroup)
                        return Response(ResponseStatus.SUCCESS, "Элемент успешно удалён!\n")
                    } else return Response(
                        ResponseStatus.ERROR,
                        "Произошла ошибка при удалении элемента из базы данных!"
                    )
                } finally {
                    locker.writeLock().unlock()
                }
            } else return Response(ResponseStatus.ERROR, "Недостаточно прав для удаления объекта.\n")
        }
        return Response(ResponseStatus.ERROR, "Элемент c id=$id не найден.\n")
    }

    /**
     * Обновляет элемент из коллекции по указанному идентификатору.
     * @param id идентификатор элемента для обновления.
     */
    fun update(id: Long): Response {
        if (dataBaseManager.exist(id)) {
            if (dataBaseManager.checkCreatorId(id)) {
                threadPoolManager.responseSenderExecutor.submit {
                    Server.sendResponse(Response(ResponseStatus.ASK_OBJECT, "Введите обновлённые данные объекта!"))
                }
                val studyGroup = threadPoolManager.objectQueue.take()
                locker.writeLock().lock()
                try {
                    if (dataBaseManager.update(id, studyGroup)) {
                        val deleteGroup =
                            collection.stream().filter { group: StudyGroup -> group.id == id }.findFirst()
                                .get()
                        collection.remove(deleteGroup)
                        HistoryWriter.writeObject(studyGroup)
                        studyGroup.id = id
                        collection.add(studyGroup)
                        return Response(ResponseStatus.SUCCESS, "Элемент успешно обновлён!\n")
                    } else return Response(
                        ResponseStatus.ERROR,
                        "Произошла ошибка при обновлении элемента в базе данных!"
                    )
                } finally {
                    locker.writeLock().unlock()
                }
            } else return Response(ResponseStatus.ERROR, "Недостаточно прав для обновления объекта.")
        } else
          return Response(ResponseStatus.ERROR, "Элемент c id=$id не найден.\n")
    }

    /**
     * Обновляет элемент из коллекции по указанному идентификатору и берёт данные для обновления из файла.
     * @param id идентификатор элемента для удаления.
     * @param reader для чтения данных из файла.
     */
    fun update(reader: BufferedReader, id: Long): Response {
        if (dataBaseManager.exist(id)) {
            if (dataBaseManager.checkCreatorId(id)) {
                val studyGroup = ObjectBuilder.create(reader)
                locker.writeLock().lock()
                try {
                    if (dataBaseManager.update(id, studyGroup)) {
                        val deleteGroup =
                            collection.stream().filter { group: StudyGroup -> group.id == id }.findFirst()
                                .get()
                        collection.remove(deleteGroup)
                        HistoryWriter.writeObject(studyGroup)
                        studyGroup.id = id
                        collection.add(studyGroup)
                        return Response(ResponseStatus.SUCCESS, "Элемент успешно обновлён!\n")
                    } else return Response(
                        ResponseStatus.ERROR,
                        "Произошла ошибка при обновлении элемента в базе данных!"
                    )
                } finally {
                    locker.writeLock().unlock()
                }
            } else return Response(ResponseStatus.ERROR, "Недостаточно прав для обновления объекта.")
        } else return Response(ResponseStatus.ERROR, "Элемент c id=$id не найден.\n")
    }

    /**
     * Удаляет все элементы, превышающие заданный.
     */
    fun removeGreater(): Response {
        if (collection.isEmpty()) return Response(ResponseStatus.SUCCESS, "Коллекция пуста!")
        threadPoolManager.responseSenderExecutor.submit {
            Server.sendResponse(Response(ResponseStatus.ASK_OBJECT, "Введите обновлённые данные объекта!"))
        }
        val compareStudyGroup = threadPoolManager.objectQueue.take()
        HistoryWriter.writeObject(compareStudyGroup)
        compareStudyGroup.id = 0
        locker.writeLock().lock()
        try {
        for (studyGroup in collection) {
            if (studyGroup > compareStudyGroup) {
                if (dataBaseManager.remove(studyGroup.id)) {
                    collection.remove(studyGroup)
                }
            }
        }
        } finally {
            locker.writeLock().unlock()
        }
        return Response(ResponseStatus.SUCCESS, "Элемент(ы) успешно удален(ы)!")
    }

    /**
     * Удаляет все элементы, превышающие тот, чьи данные берёт из файла.
     * @param reader для чтения данных из файла.
     */
    fun removeGreater(reader: BufferedReader): Response {
        if (collection.isEmpty()) return Response(ResponseStatus.INTERMEDIATE_SUCCESS, "Коллекция пуста!")
        val compareStudyGroup = ObjectBuilder.create(reader)

        HistoryWriter.writeObject(compareStudyGroup)
        compareStudyGroup.id = 0
        locker.readLock().lock()
        try {
            for (studyGroup in collection) {
                if (studyGroup > compareStudyGroup) {
                    if (dataBaseManager.checkCreatorId(studyGroup.id)) {
                        if (dataBaseManager.remove(studyGroup.id)) {
                            collection.remove(studyGroup)
                        }
                    }
                }
            }
        } finally {
            locker.readLock().unlock()
        }
        return Response(ResponseStatus.INTERMEDIATE_SUCCESS, "Элемент(ы) успешно удален(ы)!")
    }

    /**
     * Выводит количество элементов, значение поля studentsCount которых меньше заданного.
     * @param studentsCount количество студентов.
     */
    fun countLess(studentsCount: Long): Response {
        locker.readLock().lock()
        try {
            return Response(
                ResponseStatus.SUCCESS,
                collection.stream().filter { studyGroup: StudyGroup -> studyGroup.studentsCount < studentsCount }
                    .count().toString() + "\n")
        } finally {
            locker.readLock().unlock()
        }
    }

    /**
     * Выводит элементы коллекции в порядке убывания.
     */
    fun printDescending(): Response {
        locker.readLock().lock()
        try {
            if (collection.isEmpty()) return Response(ResponseStatus.SUCCESS, "Коллекция пуста!")
            collection.stream().sorted(Comparator.reverseOrder()).forEach { studyGroup: StudyGroup ->
                val future = threadPoolManager.responseSenderExecutor.submit {
                    Server.sendResponse(Response(ResponseStatus.INTERMEDIATE_SUCCESS, studyGroup.toString()))
                }
                future.get()
            }
        } finally {
            locker.readLock().unlock()
        }
        return Response(ResponseStatus.SUCCESS, "")
    }

    /**
     * Выводит значения поля formOfEducation всех элементов в порядке возрастания.
     */
    fun printFieldAscending(): Response {
        locker.readLock().lock()
        try {
            if (collection.isEmpty()) return Response(ResponseStatus.SUCCESS, "Коллекция пуста!")
            collection.stream().sorted().forEach { studyGroup: StudyGroup ->
                val future = threadPoolManager.responseSenderExecutor.submit {
                    Server.sendResponse(Response(ResponseStatus.INTERMEDIATE_SUCCESS, studyGroup.formOfEducation.toString()))
                }
                future.get()
            }
            return Response(ResponseStatus.SUCCESS, "")
        } finally {
            locker.readLock().unlock()
        }
    }

    /**
     * Выводит информацию о коллекции.
     */
    fun info(): Response {
        locker.readLock().lock()
        try {
            val info = """
                 Тип коллекции: PriorityQueue
                 Тип элементов: StudyGroup
                 Количество элементов текущего пользователя: ${dataBaseManager.countUsersObjects()}
                 Количество всех элементов: ${collection.size}
                 Дата инициализации: ${initializationTime.format(DateTimeFormatter.ofPattern("dd.MM.yyyy"))}
                 """.trimIndent()
            return Response(ResponseStatus.SUCCESS, info + "\n")
        } finally {
            locker.readLock().unlock()
        }
    }
}
